File: archive/cdk-ts/Pr483/lib/iam-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

export interface IamStackProps extends cdk.StackProps {
  environmentSuffix?: string;
}

export class IamStack extends cdk.Stack {
  public readonly ec2Role: iam.Role;
  public readonly instanceProfile: iam.InstanceProfile;

  constructor(scope: Construct, id: string, props?: IamStackProps) {
    super(scope, id, props);

    const environmentSuffix = props?.environmentSuffix || 'dev';

    // Create IAM role for EC2 instances
    this.ec2Role = new iam.Role(this, `EC2Role${environmentSuffix}`, {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
      description: `IAM role for EC2 instances in ${environmentSuffix} environment`,
      roleName: `EC2-MultiTier-Role-${environmentSuffix}`,
    });

    // Add policies for AWS services interaction
    this.ec2Role.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore')
    );

    // Add custom policy for additional AWS services
    this.ec2Role.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: [
          's3:GetObject',
          's3:PutObject',
          's3:DeleteObject',
          's3:ListBucket',
          'cloudwatch:PutMetricData',
          'logs:CreateLogGroup',
          'logs:CreateLogStream',
          'logs:PutLogEvents',
          'logs:DescribeLogStreams',
        ],
        resources: ['*'],
      })
    );

    // Create instance profile
    this.instanceProfile = new iam.InstanceProfile(
      this,
      `InstanceProfile${environmentSuffix}`,
      {
        role: this.ec2Role,
        instanceProfileName: `EC2-MultiTier-Profile-${environmentSuffix}`,
      }
    );

    // Outputs
    new cdk.CfnOutput(this, 'EC2RoleArn', {
      value: this.ec2Role.roleArn,
      description: 'EC2 IAM Role ARN',
    });

    new cdk.CfnOutput(this, 'InstanceProfileArn', {
      value: this.instanceProfile.instanceProfileArn,
      description: 'Instance Profile ARN',
    });
  }
}

File: archive/cdk-java/Pr2283/lib/src/main/java/app/IamStack.java
package app;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.iam.*;
import software.amazon.awscdk.services.s3.Bucket;
import software.constructs.Construct;

import java.util.Map;

public class IamStack extends Stack {

    private final Role ec2Role;

    public IamStack(final Construct scope, final String id, 
                   final StackProps props, final Bucket s3Bucket) {
        super(scope, id, props);

        // Create IAM role for EC2 with S3 read-only access
        this.ec2Role = Role.Builder.create(this, "app-role-ec2")
                .roleName("app-role-ec2")
                .description("EC2 role with S3 read-only access")
                .assumedBy(new ServicePrincipal("ec2.amazonaws.com"))
                .build();

        // Add policy for S3 read-only access to specific bucket
        PolicyStatement s3ReadPolicy = PolicyStatement.Builder.create()
                .effect(Effect.ALLOW)
                .actions(java.util.List.of(
                        "s3:GetObject",
                        "s3:GetObjectVersion",
                        "s3:ListBucket"
                ))
                .resources(java.util.List.of(
                        s3Bucket.getBucketArn(),
                        s3Bucket.getBucketArn() + "/*"
                ))
                .build();

        ec2Role.addToPolicy(s3ReadPolicy);

        this.addCommonTags();
    }

    private void addCommonTags() {
        Map<String, String> tags = Map.of(
                "Project", "CloudSecurity",
                "Environment", "Production"
        );
        tags.forEach((key, value) -> this.getNode().addMetadata(key, value));
    }

    public Role getEc2Role() {
        return ec2Role;
    }
}